### PR TITLE
Linting with typescript

### PR DESCRIPTION
**Pull request by vankop**
https://github.com/webpack/enhanced-resolve/issues/184

**What kind of change does this PR introduce?**

typescript linting for source code and tests

**Did you add tests for your changes?**

No

**Does this PR introduce a breaking change?**

No 

**What needs to be documented once your changes are merged?**

Nothing

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +30,8 @@
     "mocha": "^6.1.4",
     "nyc": "^14.1.1",
     "prettier": "^1.15.2",
-    "should": "^8.0.2"
+    "should": "^13.2.3",
```
Comment (vankop) [inline package.json]: I have upgraded should.js to latest version where typescript definitions are presented

File: test/forEachBail.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ describe("forEachBail", () => {
 		forEachBail(
 			[],
 			() => {
-				done.fail();
+				done("fail");
```
Comment (vankop) [inline test/forEachBail.js]: looks like `Done` accept error as first argument and have not `fail` method

File: lib/ResolverFactory.js [hunk 3]
```diff
@@ -64,13 +64,15 @@ exports.createResolver = function(options) {
 	let alias = options.alias || [];
 
 	// A PnP API that should be used - null is "never", undefined is "auto"
-	const pnpApi =
-		options.pnpApi === undefined
-			? process.versions.pnp
-				? // eslint-disable-next-line node/no-missing-require
-				  require("pnpapi")
```
Comment (vankop) [inline lib/ResolverFactory.js]: can not place `@ts-ignore` in ternary operator because of prettier errors

File: package.json [hunk 4]
```diff
@@ -37,12 +40,13 @@
...
   "scripts": {
     "lint": "eslint lib test",
+    "types": "tsc",
     "pretty": "prettier --loglevel warn --write \"{lib,test}/**/*.{js,json}\"",
     "pretest": "yarn lint",
     "test": "mocha --full-trace --check-leaks",
     "precover": "yarn lint",
     "cover": "nyc --reporter=html node node_modules/mocha/bin/_mocha",
-    "pretravis": "yarn lint",
+    "pretravis": "yarn lint && yarn types",
```
Comment (sokra) [inline package.json]:
```suggestion
    "pretravis": "yarn lint",
```

File: package.json [hunk 5]
```diff
@@ -37,12 +40,13 @@
   "homepage": "http://github.com/webpack/enhanced-resolve",
   "scripts": {
     "lint": "eslint lib test",
```
Comment (sokra) [inline package.json]:
```suggestion
     "lint": "eslint lib test && tsc",
```

File: test/forEachBail.js (hunk 2 above)
Comment (sokra) [inline test/forEachBail.js]:
```suggestion
				done(new Error("Should not be called");
```

File: test/missing.js [hunk 6]
```diff
@@ -74,7 +74,9 @@ describe("missing", function() {
 				var callback = function(err, filename) {
 					Array.from(missingDependencies)
 						.sort()
-						.should.containDeep(testCase[2].sort());
+						.should.containDeep(
+							/** @type {Array<string>} */ (testCase[2]).sort()
```
Comment (sokra) [inline test/missing.js]: You can type `testCases` as `[string, string, string[]][]`

File: package.json (hunk 4 above)
Comment (vankop) [reply package.json]: done

File: package.json (hunk 5 above)
Comment (vankop) [reply package.json]: done

File: test/forEachBail.js (hunk 2 above)
Comment (vankop) [reply test/forEachBail.js]: done

File: test/missing.js (hunk 6 above)
Comment (vankop) [reply test/missing.js]: done

**Comment by sokra** [comment]
Thanks

**Pull request merged**